### PR TITLE
Pc 6pm 4 custom google api

### PR DESCRIPTION
**Pull request by pconrad**
This PR started with work done by W21 team-6pm-4 as part of PR #191 (closed, subsumed by this one), and continued with refactoring done by @pconrad.

# Original work from #191
 
In this PR, we:
- Added objects in App User to store Google API tokens
- Added check to make sure API tokens are valid
- User will be able to use their own API token to search. They can also reset to the default token by attempting to submit any random invalid token. 
- This PR resolves [135](https://github.com/ucsb-cs156-w21/proj-mapache-search/issues/135)

capture1

# Additional work by @pconrad subsequent to #191

- separate accounting of user-based vs. global API search token, and 
- add a `Counter` table for persisting global counters.  Used right now only for tracking daily usage of global search api, but may be useful for other purposes as well.
- refactor Google search code for improved ability to test (code was excluded from test coverage previously at the `pom.xml` level in configuration of the `jacoco` plugin.)
- use lombok to simplify a few "data" classes (this work can continue during S21)

### CODE REVIEW

File: src/main/java/edu/ucsb/mapache/google/SearchResult.java [hunk 1]
```diff
@@ -99,6 +49,35 @@ public static SearchResult fromJSON(String json) {
...
+     * Create a SearchResult object from json representation
+     * 
+     * @param json String of json returned by Google Search API
+     * @return a new SearchResult object
+     */
+    public String toJSON() throws JsonProcessingException {
+        ObjectMapper objectMapper = new ObjectMapper();
+        // try {
+        return objectMapper.writeValueAsString(this);
+        // } catch (JsonProcessingException jpe) {
```
Comment (JacquelineMai) [inline src/main/java/edu/ucsb/mapache/google/SearchResult.java:65]: Should this commented try/catch be removed?

File: src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java [hunk 2]
```diff
@@ -21,405 +19,251 @@
...
+        int index_trim = retval.indexOf("\"items\"");
+        retval = retval.substring(0, index_trim) + "\"items\": []}";
+        when(googleSearchServiceHelper.getJSON(any(), any())).thenReturn(retval);
+        SlackSlashCommandParams params = new SlackSlashCommandParams();
+        params.setText("search google test");
+        RichMessage result = slackSlashCommandController.googleSearch(params);
+        assertEquals("No results found!", result.getAttachments()[0].getText());
+    }
+
+    // WIP unable to test IOexception currently
```
Comment (JacquelineMai) [inline src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java:194]: Is this still the case?

File: src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java [hunk 3]
```diff
@@ -21,405 +19,251 @@
...
+    @Test
+    public void test_googleSearch_1() throws Exception {
```
Comment (JacquelineMai) [inline src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java:196]: `test_googleSearch_1` through `4` could use more informative names that the specific cases each test is looking at, for documentation purposes

File: src/test/java/edu/ucsb/mapache/entities/SearchTests.java [hunk 4]
```diff
@@ -8,71 +8,81 @@
...
-    Search s = new Search(1L, "term", 1);
-    String expected = "{ id='1', searchTerm='term', count='1'}";
-    assertEquals(expected, s.toString());
-  }
+  // @Test
+  // public void test_toString() throws Exception {
+  //   Search s = new Search(1L, "term", 1);
+  //   String expected = "{ id='1', searchTerm='term', count='1'}";
+  //   assertEquals(expected, s.toString());
+  // }
```
Comment (JacquelineMai) [inline src/test/java/edu/ucsb/mapache/entities/SearchTests.java:86]: Do we need these commented tests, or can they be removed?

File: src/test/java/edu/ucsb/mapache/services/CounterServiceTests.java [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        // Short version: it captures the value that was passed to the save method
+
+        final ArgumentCaptor<Counter> captor = ArgumentCaptor.forClass(Counter.class);
+        verify(mockCounterRepository).save(captor.capture());
+        final Counter savedCounter = captor.getValue();
+
+        Counter cExpected = new Counter("foo", 99);
+        cExpected.setLastReset(lessThanADayAgo);
+        assertEquals(cExpected, cReturned);
+        // assertEquals(cExpected, savedCounter);
```
Comment (JacquelineMai) [inline src/test/java/edu/ucsb/mapache/services/CounterServiceTests.java:189]: This commented line can be removed

**Review by JacquelineMai — CHANGES_REQUESTED** [review]

File: src/test/java/edu/ucsb/mapache/entities/SearchTests.java (hunk 4 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/mapache/entities/SearchTests.java:86]: Ah.. good catch.  They can be removed because of Lombok I think... I'll take a look..

Comment (pconrad) [reply src/test/java/edu/ucsb/mapache/entities/SearchTests.java]: Removed

File: src/test/java/edu/ucsb/mapache/services/CounterServiceTests.java (hunk 5 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/mapache/services/CounterServiceTests.java:189]: Agreed... done

File: src/main/java/edu/ucsb/mapache/google/SearchResult.java (hunk 1 above)
Comment (pconrad) [reply src/main/java/edu/ucsb/mapache/google/SearchResult.java]: Good catch... done!

File: src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java (hunk 3 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java]: Agreed... done.   

New names reflect the main difference among the four tests (i.e. passing different values for the command params after `/mapache`)

File: src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java (hunk 2 above)
Comment (pconrad) [reply src/test/java/edu/ucsb/mapache/controllers/SlackSlashCommandControllerTests.java]: I don't think so.  So I removed that comment.

**Review by JacquelineMai — APPROVED** [review]
lgtm

**Review by andrewhlu — APPROVED** [review]

**Pull request merged**